### PR TITLE
add ia3 and adalora support

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
https://github.com/huggingface/peft/pull/1540 fixed adalora rank==0 issue in inference load model.

**Comment by sywangyi** [comment]
update peft to 0.10.0 which contains https://github.com/huggingface/peft/pull/1540

**Comment by sywangyi** [comment]
adalora + deepspeed zero3 finetune fix in peft https://github.com/huggingface/peft/pull/1625

**Comment by sywangyi** [comment]
IA3 finetune with "--use_flash_attention" need PR fix in https://github.com/huggingface/peft/pull/1634. since q, k, v should has some data type in FusedSDPA.apply

**Review by regisss — COMMENTED** [review]
For PEFT v0.10, there was an issue with higher memory consumption from PEFT v0.7: https://github.com/huggingface/optimum-habana/issues/590
Let me try to see if that works.

File: examples/language-modeling/run_lora_clm.py [hunk 1]
```diff
@@ -289,12 +289,53 @@ class FinetuneArguments:
...
+        },
+    )
+    adalora_delta_t: int = field(
+        default=10,
+        metadata={"help": "Interval of steps for AdaLoRA to update rank"},
+    )
+    adalora_orth_reg_weight: float = field(
+        default=0.5,
+        metadata={"help": "Orthogonal regularization weight for AdaLoRA"},
+    )
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/language-modeling/run_lora_clm.py:323]: yes

File: optimum/habana/peft/layer.py [hunk 2]
```diff
@@ -29,3 +29,33 @@ def GaudiLoraLayerLinearForward(self, x: torch.Tensor, *args: Any, **kwargs: Any
...
+                continue
+            lora_A = self.lora_A[active_adapter]
+            lora_B = self.lora_B[active_adapter]
+            lora_E = self.lora_E[active_adapter]
+            dropout = self.lora_dropout[active_adapter]
+            scaling = self.scaling[active_adapter]
+            ranknum = self.ranknum[active_adapter] + 1e-5
+
+            x = x.to(lora_A.dtype)
+            result += (dropout(x) @ (lora_A * lora_E).T @ lora_B.T) * (scaling / ranknum)
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply optimum/habana/peft/layer.py:59]: yes.

**Comment by regisss** [comment]
We're still having an issue with Falcon-180b and more recent versions of PEFT. Let me see how we can do.

**Comment by sywangyi** [comment]
try https://github.com/huggingface/optimum-habana/pull/895 and it could fix falcon 180b issue

**Comment by regisss** [comment]
@sywangyi LGTM! Can you share a command for both ia3 and adalora to make sure I can run them please?

**Comment by sywangyi** [comment]
Hi, @regisss . you could use "--peft_type ia3" for IA3 and "--peft_type adalora" for adalora in run_lora_clm.py command for alpaca dataset

like 
python3 run_lora_clm.py \
  --model_name_or_path meta-llama/Llama-2-7b-hf \
  --dataset_name tatsu-lab/alpaca \
  --output_dir ./lora_out \
  --num_train_epochs 1 \
  --max_step 500 \
  --max_seq_len 2048 \
  --per_device_train_batch_size 1 \
  --per_device_eval_batch_size 10 \
  --evaluation_strategy epoch \
  --eval_delay 2 \
  --save_strategy no \
  --learning_rate 0.0018 \
  --warmup_ratio 0.03 \
  --lr_scheduler_type "cosine" \
  --logging_steps 1 \
  --dataset_concatenation \
  --do_train \
  --do_eval \
  --use_habana \
  --use_lazy_mode \
  --throughput_warmup_steps 3 \
  --lora_rank 4 \
  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
  --validation_split_percentage 4 \
  --peft_type ia3 \
  --pipelining_fwd_bwd \
  --bf16 \
  --attn_softmax_bf16 True 

python3 run_lora_clm.py \
  --model_name_or_path meta-llama/Llama-2-7b-hf \
  --dataset_name tatsu-lab/alpaca \
  --output_dir ./lora_out \
  --num_train_epochs 1 \
  --max_step 500 \
  --max_seq_len 2048 \
  --per_device_train_batch_size 1 \
  --per_device_eval_batch_size 10 \
  --evaluation_strategy epoch \
  --eval_delay 2 \
  --save_strategy no \
  --learning_rate 0.0018 \
  --warmup_ratio 0.03 \
  --lr_scheduler_type "cosine" \
  --logging_steps 1 \
  --dataset_concatenation \
  --do_train \
  --do_eval \
  --use_habana \
  --use_lazy_mode \
  --throughput_warmup_steps 3 \
  --lora_rank 4 \
  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
  --validation_split_percentage 4 \
  --peft_type adalora\
  --pipelining_fwd_bwd \
  --bf16 \
  --attn_softmax_bf16 True

**Review by regisss — APPROVED** [review]

**Pull request merged**